executor/src/witgen/machines/double_sorted_witness_machine.rs

```diff
@@ -1,199 +1,58 @@
     ) -> Option<Self> {
-        let degree_range = parts.common_degree_range();
-
-        let degree = degree_range.max;
-
-        // get the namespaces and column names
-        let (mut namespaces, columns): (HashSet<_>, HashSet<_>) = parts
-            .witnesses
-            .iter()
-            .map(|r| split_column_name(parts.column_name(r)))
-            .unzip();
-
-        if namespaces.len() > 1 {
-            // columns are not in the same namespace, fail
-            return None;
-        }
-
-        if !parts
-            .connecting_identities
-            .values()
-            .all(|i| i.kind == IdentityKind::Permutation)
-        {
-            return None;
-        }
-
-        let selector_ids = parts
-            .connecting_identities
-            .values()
-            .map(|i| {
-                i.right
-                    .selector
-                    .as_ref()
-                    .and_then(|r| try_to_simple_poly(r))
-                    .map(|p| (i.id, p.poly_id))
-            })
-            .collect::<Option<BTreeMap<_, _>>>()?;
-
-        let namespace = namespaces.drain().next().unwrap().into();
-
-        // TODO check the identities.
-        let selector_names = selector_ids
-            .values()
-            .map(|s| split_column_name(parts.column_name(s)).1);
-        let allowed_witnesses: HashSet<_> = ALLOWED_WITNESSES
-            .into_iter()
-            .chain(selector_names)
-            .collect();
-        if !columns.iter().all(|c| allowed_witnesses.contains(c)) {
-            return None;
-        }
-
-        let has_diff_columns = DIFF_COLUMNS.iter().all(|c| columns.contains(c));
-        let has_bootloader_write_column = columns.contains(&BOOTLOADER_WRITE_COLUMN);
-
-        let diff_columns_base = if has_diff_columns {
-            // We have the `m_diff_upper` and `m_diff_lower` columns.
-            // Now, we check that they both have the same range constraint and use it to determine
-            // the base of the two digits.
-            let upper_poly_id =
-                fixed_data.try_column_by_name(&format!("{namespace}::{}", DIFF_COLUMNS[0]))?;
-            let upper_range_constraint = fixed_data.global_range_constraints().witness_constraints
-                [&upper_poly_id]
-                .as_ref()?;
-            let lower_poly_id =
-                fixed_data.try_column_by_name(&format!("{namespace}::{}", DIFF_COLUMNS[1]))?;
-            let lower_range_constraint = fixed_data.global_range_constraints().witness_constraints
-                [&lower_poly_id]
-                .as_ref()?;
-
-            let (min, max) = upper_range_constraint.range();
-
-            if upper_range_constraint == lower_range_constraint && min == T::zero() {
-                Some(max.to_degree() + 1)
-            } else {
-                return None;
+        match T::known_field().unwrap() {
+            KnownField::BabyBearField | KnownField::Mersenne31Field => {
```

Comment:
I don't think we should make it dependent on the field. This way, we can't have the 16-Bit version on Bn254. Which is unfortunate, because `halo2-mock` happens to have the best error messages if a constraint is failing!

Reply:
I'm fine with either. The only thing is that if we don't make it dependent on `T`, we have to pass a new argument around, not sure how much that's gonna mess up/make things uglier.